src/compiler/transformers/automatic-key-insertion/index.ts

```diff
@@ -0,0 +1,136 @@
+   * @returns the result of handling the node
+   */
+  function findClassDeclVisitor(node: ts.Node): ts.VisitResult<ts.Node> {
+    if (ts.isClassDeclaration(node)) {
+      return ts.visitEachChild(node, findRenderMethodVisitor, transformCtx);
+    } else {
+      return ts.visitEachChild(node, findClassDeclVisitor, transformCtx);
+    }
+  }
+
+  /**
+   * This middle visitor function is responsible for finding the render method
+   * on a Stencil class and then passing off responsibility to the inner-most
+   * visitor, which deals with syntax nodes inside the method.
+   *
+   * @param node a typescript syntax tree node
+   * @returns the result of handling the node
+   */
+  function findRenderMethodVisitor(node: ts.Node): ts.VisitResult<ts.Node> {
+    if (ts.isMethodDeclaration(node) && node.name.getText() === 'render') {
+      return ts.visitEachChild(node, jsxElementVisitor, transformCtx);
+    } else {
+      return ts.visitEachChild(node, findRenderMethodVisitor, transformCtx);
+    }
+  }
+
+  /**
+   * Our inner-most visitor function. This will edit any JSX nodes that it
+   * finds, adding a `key` attribute to them via {@link addKeyAttr}.
+   *
+   * @param node a typescript syntax tree node
+   * @returns the result of handling the node
+   */
+  function jsxElementVisitor(node: ts.Node): ts.VisitResult<ts.Node> {
+    if (isJSXElWithAttrs(node)) {
+      return addKeyAttr(node);
+    } else {
+      return ts.visitEachChild(node, jsxElementVisitor, transformCtx);
+    }
+  }
+
+  return (tsSourceFile) => {
+    return ts.visitEachChild(tsSourceFile, findClassDeclVisitor, transformCtx);
+  };
+};
+
+/**
+ * Type guard to see if a TypeScript syntax node is one of the node types which
+ * corresponds to a JSX element that can have attributes on it. This is either
+ * an opening node, like `<div attr="hey">`, or a 'self-closing' node like
+ * `<i class="best" />`.
+ *
+ * @param node a typescript syntax tree node
+ * @returns whether or not the node is JSX node which could have attributes
+ */
+function isJSXElWithAttrs(node: ts.Node): node is ts.JsxOpeningElement | ts.JsxSelfClosingElement {
+  return ts.isJsxOpeningElement(node) || ts.isJsxSelfClosingElement(node);
+}
+
+/**
+ * Given a JSX syntax tree node update it to include a unique key attribute.
+ * This will respect any attributes already set on the node, including a
+ * pre-existing, user-defined `key` attribute.
+ *
+ * @param jsxElement a typescript JSX syntax tree node
+ * @returns an updated JSX element, with a key added.
+ */
+function addKeyAttr(
+  jsxElement: ts.JsxOpeningElement | ts.JsxSelfClosingElement,
+): ts.JsxOpeningElement | ts.JsxSelfClosingElement {
+  const updatedAttributes = ts.factory.createJsxAttributes([
+    ts.factory.createJsxAttribute(
+      ts.factory.createIdentifier('key'),
+      ts.factory.createStringLiteral(deriveJSXKey(jsxElement)),
+    ),
+    ...jsxElement.attributes.properties,
+  ]);
```

Comment:
it won't because we splat in the existing attrs after the new one

Reply:
ah so this actually works more strangely that one might think!

I think TypeScript is perfectly fine with you doing

```tsx
<foo-bar key="first" key="second" />
```

because what it means to have multiple keys with the same name is really a runtime question

I'll add a check to ensure there's not already a property with the name `"key"` before we add

Reply:
I added that check and also a test that if you've already set a `key` attr we won't add another one